src/script/src/python/engine.rs

```diff
@@ -36,11 +43,117 @@ use crate::python::error::{self, Result};
 
 const PY_ENGINE: &str = "python";
 
+pub struct PyUdf {
+    copr: CoprocessorRef,
+}
+
+impl std::fmt::Display for PyUdf {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", &self.copr.name)
+    }
+}
+
+impl PyUdf {
+    fn from_copr(copr: CoprocessorRef) -> Arc<Self> {
+        Arc::new(Self { copr })
+    }
+
+    /// Register to `FUNCTION_REGISTRY`
+    fn register_as_udf(zelf: Arc<Self>) {
+        FUNCTION_REGISTRY.register(zelf)
+    }
+    fn register_to_query_engine(zelf: Arc<Self>, engine: QueryEngineRef) {
+        engine.register_function(zelf)
+    }
+
+    /// Fake a schema, should only be used with dynamically eval a Python Udf
+    fn fake_schema(&self, columns: &[VectorRef]) -> SchemaRef {
+        let arg_names = &self.copr.deco_args.arg_names;
+        let col_sch: Vec<_> = columns
+            .iter()
+            .enumerate()
+            .map(|(i, col)| ColumnSchema::new(arg_names[i].to_owned(), col.data_type(), true))
+            .collect();
+        let schema = datatypes::schema::Schema::new(col_sch);
+        Arc::new(schema)
+    }
+}
+
+impl Function for PyUdf {
+    fn name(&self) -> &str {
+        &self.copr.name
+    }
+
+    fn return_type(
+        &self,
+        _input_types: &[datatypes::prelude::ConcreteDataType],
+    ) -> common_query::error::Result<datatypes::prelude::ConcreteDataType> {
+        // TODO: use correct return annotation if exist
+        Ok(self.copr.return_types[0]
+            .clone()
+            .map_or(ConcreteDataType::float64_datatype(), |anno| {
+                anno.datatype
+                    .unwrap_or(ConcreteDataType::float64_datatype())
+            }))
+    }
+
+    fn signature(&self) -> common_query::prelude::Signature {
+        Signature::uniform(
+            self.copr.arg_types.len(),
+            ConcreteDataType::numerics(),
```

Comment:
Just accepts numeric types? Why not use the actual types in `copr.arg_types`.

Reply:
> Just accepts numeric types? Why not use the actual types in `copr.arg_types`.

Now it's Signature::any unless all input args have known type